File: ios/App/ZenWidgets/WidgetSnapshot.swift
import Foundation

/// Mirror of src/bridge/widget-snapshot.ts — the app is the writer. Every
/// field a later shell might add or drop decodes as optional, so an older
/// extension never fails on a newer snapshot (and vice versa).
struct WidgetThemeData: Decodable {
    var mode: String?
    var bg: String?
    var bg1: String?
    var bg2: String?
    var fg: String?
    var fg2: String?
    var muted: String?
    var accent: String?
    var red: String?
}

struct WidgetNote: Decodable, Identifiable {
    var id: String { path }
    let path: String
    let title: String
    let folder: String?
    /// ms since epoch.
    let updatedAt: Double
    let pinned: Bool

    var updatedDate: Date { Date(timeIntervalSince1970: updatedAt / 1000) }
}

struct WidgetTask: Decodable, Identifiable {
    let id: String
    let path: String
    let noteTitle: String
    let content: String
    /// ISO YYYY-MM-DD; nil for an undated task.
    let due: String?
    let overdue: Bool
    let inProgress: Bool
    let priority: String?
}

struct WidgetTaskCounts: Decodable {
    let today: Int
    let overdue: Int
}

struct WidgetSnapshot: Decodable {
    let version: Int
    /// ms since epoch.
    let generatedAt: Double
    let vaultName: String?
    let theme: WidgetThemeData?
    let notes: [WidgetNote]
    let tasks: [WidgetTask]
    let taskCounts: WidgetTaskCounts?
    let tasksReady: Bool?

    var generatedDate: Date { Date(timeIntervalSince1970: generatedAt / 1000) }
    var todayCount: Int { taskCounts?.today ?? tasks.count }
    var overdueCount: Int { taskCounts?.overdue ?? tasks.filter { $0.overdue }.count }
}

enum WidgetSnapshotStore {
    static let appGroupId = "group.md.zennotes"
    /// Mirrored by WidgetBridgePlugin.snapshotPath in the app.
    static let relativePath = "widgets/snapshot.json"

    static var url: URL? {
        FileManager.default
            .containerURL(forSecurityApplicationGroupIdentifier: appGroupId)?
            .appendingPathComponent(relativePath)
    }

    static func load() -> WidgetSnapshot? {
        guard let url = url, let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(WidgetSnapshot.self, from: data)
    }

    /// Gallery previews and placeholders — what a lived-in vault looks like.
    static func sample(now: Date = Date()) -> WidgetSnapshot {
        let ms = now.timeIntervalSince1970 * 1000
        let minute = 60_000.0
        let hour = 3_600_000.0
        let today = WidgetFormat.isoDate(now)
        let yesterday = WidgetFormat.isoDate(now.addingTimeInterval(-86_400))
        return WidgetSnapshot(
            version: 1,
            generatedAt: ms,
            vaultName: "My Vault",
            theme: nil,
            notes: [
                WidgetNote(path: "inbox/Reading list.md", title: "Reading list", folder: "inbox",
                           updatedAt: ms - 25 * minute, pinned: true),
                WidgetNote(path: "inbox/Product ideas.md", title: "Product ideas", folder: "inbox",
                           updatedAt: ms - 3 * hour, pinned: false),
                WidgetNote(path: "inbox/Meeting notes.md", title: "Meeting notes", folder: "inbox",
                           updatedAt: ms - 6 * hour, pinned: false),
                WidgetNote(path: "quick/Grocery run.md", title: "Grocery run", folder: "quick",
                           updatedAt: ms - 26 * hour, pinned: false),
                WidgetNote(path: "inbox/Trip planning.md", title: "Trip planning", folder: "inbox",
                           updatedAt: ms - 2 * 24 * hour, pinned: false),
                WidgetNote(path: "inbox/Weekly review.md", title: "Weekly review", folder: "inbox",
                           updatedAt: ms - 3 * 24 * hour, pinned: false),
                WidgetNote(path: "inbox/Book notes.md", title: "Book notes", folder: "inbox",
                           updatedAt: ms - 4 * 24 * hour, pinned: false),
                WidgetNote(path: "inbox/Recipes.md", title: "Recipes", folder: "inbox",
                           updatedAt: ms - 5 * 24 * hour, pinned: false),
                WidgetNote(path: "inbox/Journal.md", title: "Journal", folder: "inbox",
                           updatedAt: ms - 6 * 24 * hour, pinned: false)
            ],
            tasks: [
                WidgetTask(id: "inbox/Today.md#0", path: "inbox/Today.md", noteTitle: "Today",
                           content: "Reply to the design review", due: today, overdue: false,
                           inProgress: false, priority: nil),
                WidgetTask(id: "inbox/Trip planning.md#2", path: "inbox/Trip planning.md",
                           noteTitle: "Trip planning", content: "Book the October flights",
                           due: yesterday, overdue: true, inProgress: false, priority: "high"),
                WidgetTask(id: "inbox/Product ideas.md#1", path: "inbox/Product ideas.md",
                           noteTitle: "Product ideas", content: "Draft the release notes",
                           due: today, overdue: false, inProgress: true, priority: nil),
                WidgetTask(id: "inbox/Today.md#3", path: "inbox/Today.md", noteTitle: "Today",
                           content: "Call the dentist", due: nil, overdue: false,
                           inProgress: false, priority: nil),
                WidgetTask(id: "inbox/Weekly review.md#0", path: "inbox/Weekly review.md",
                           noteTitle: "Weekly review", content: "Plan next week's focus",
                           due: today, overdue: false, inProgress: false, priority: nil),
                WidgetTask(id: "quick/Grocery run.md#0", path: "quick/Grocery run.md",
                           noteTitle: "Grocery run", content: "Pick up coffee beans",
                           due: nil, overdue: false, inProgress: false, priority: nil)
            ],
            taskCounts: WidgetTaskCounts(today: 6, overdue: 1),
            tasksReady: true
        )
    }
}

File: src/bridge/widget-snapshot.ts
/**
 * The Home Screen / Lock Screen widget snapshot — the contract between the
 * shell and the WidgetKit extension (ios/App/ZenWidgets).
 *
 * A widget extension runs in its own sandbox and cannot see the vault
 * (on-device vaults live in the app's Documents container, iCloud vaults in
 * the ubiquity container), so the app publishes what the widgets show — the
 * active vault's pinned + recent notes, today's tasks, and the active theme's
 * colors — as one JSON document in the App Group: WidgetBridgePlugin.swift
 * writes it, WidgetSnapshot.swift decodes it. Keep the two sides in step;
 * the Swift decoder treats every field a later shell might add as optional.
 *
 * Only pure selectors live here (node --test covers them). The store/pins
 * wiring and the native call are in widgets.ts.
 */
import type { NoteMeta } from '@zennotes/bridge-contract/ipc'
import type { VaultTask } from '@zennotes/shared-domain/tasks'
import { parseThemeBackdropColor } from './keyboard-backdrop-color.ts'

export const WIDGET_SNAPSHOT_VERSION = 1
/** Large Recent Notes shows nine rows; a few spares cover pins that vanish. */
export const WIDGET_MAX_NOTES = 12
/** Large Tasks shows eight rows plus a "+N more" footer. */
export const WIDGET_MAX_TASKS = 12

export interface WidgetTheme {
  mode: 'light' | 'dark'
  /** Hex colors (#rrggbb) sampled from the app-core `--z-*` tokens. */
  bg: string
  bg1: string
  bg2: string
  fg: string
  fg2: string
  muted: string
  accent: string
  red: string
}

export interface WidgetNote {
  path: string
  title: string
  folder: string
  /** ms since epoch, as NoteMeta reports it. */
  updatedAt: number
  pinned: boolean
}

export interface WidgetTask {
  /** VaultTask id (`${sourcePath}#${taskIndex}`) — the tap link carries it
   *  back so the shell can jump to the exact line. */
  id: string
  path: string
  noteTitle: string
  content: string
  /** ISO YYYY-MM-DD or null for an undated task (those sit in Today too). */
  due: string | null
  overdue: boolean
  inProgress: boolean
  priority: string | null
}

export interface WidgetTaskCounts {
  /** Everything in the Today bucket, not just the rows that fit. */
  today: number
  overdue: number
}

export interface WidgetSnapshot {
  version: typeof WIDGET_SNAPSHOT_VERSION
  /** ms since epoch. */
  generatedAt: number
  vaultName: string | null
  theme: WidgetTheme
  /** Pinned notes first (in pin order), then the most recently edited. */
  notes: WidgetNote[]
  /** The Today bucket, in the Tasks view's order. */
  tasks: WidgetTask[]
  taskCounts: WidgetTaskCounts
  /** False until the first task scan for this vault has landed, so the
   *  widget shows "loading" rather than a misleading "All clear". */
  tasksReady: boolean
}

/** ZenNotes' default theme (dark-hard), used before the first publish and
 *  for any token the active theme leaves undefined. */
export const FALLBACK_WIDGET_THEME: WidgetTheme = {
  mode: 'dark',
  bg: '#1d2021',
  bg1: '#32302f',
  bg2: '#3c3836',
  fg: '#d4be98',
  fg2: '#ddc7a1',
  muted: '#a89984',
  accent: '#e78a4e',
  red: '#ea6962'
}

const THEME_TOKENS: Record<Exclude<keyof WidgetTheme, 'mode'>, string> = {
  bg: '--z-bg',
  bg1: '--z-bg-1',
  bg2: '--z-bg-2',
  fg: '--z-fg',
  fg2: '--z-fg-2',
  muted: '--z-grey-2',
  accent: '--z-accent',
  red: '--z-red'
}

/** `"29 32 33"` (the `--z-*` channel triplet form) → `"#1d2021"`. */
export function channelsToHex(value: string): string | null {
  const color = parseThemeBackdropColor(value)
  if (!color) return null
  return (
    '#' +
    [color.red, color.green, color.blue].map((n) => n.toString(16).padStart(2, '0')).join('')
  )
}

/** Resolve the widget palette from a token reader (getComputedStyle in the
 *  app); tokens that don't parse keep the fallback value. */
export function themeFromTokens(
  read: (token: string) => string,
  mode: WidgetTheme['mode']
): WidgetTheme {
  const theme: WidgetTheme = { ...FALLBACK_WIDGET_THEME, mode }
  for (const [key, token] of Object.entries(THEME_TOKENS) as Array<
    [Exclude<keyof WidgetTheme, 'mode'>, string]
  >) {
    const hex = channelsToHex(read(token))
    if (hex) theme[key] = hex
  }
  return theme
}

export type WidgetNoteSource = Pick<NoteMeta, 'path' | 'title' | 'folder' | 'updatedAt'>

/**
 * Pinned notes first, in the order they were pinned (the drawer's own
 * convention — pins sort to the top of their group), then the most recently
 * edited notes, mirroring the Home dashboard's Recent list. Trash and
 * Archive never show; a pin whose note is gone is skipped, not surfaced.
 */
export function selectWidgetNotes(
  notes: readonly WidgetNoteSource[],
  pinnedPaths: readonly string[],
  max = WIDGET_MAX_NOTES
): WidgetNote[] {
  const live = notes.filter((n) => n.folder !== 'trash' && n.folder !== 'archive')
  const byPath = new Map(live.map((n) => [n.path, n] as const))
  const out: WidgetNote[] = []
  const seen = new Set<string>()
  const push = (note: WidgetNoteSource, pinned: boolean): void => {
    if (seen.has(note.path) || out.length >= max) return
    seen.add(note.path)
    out.push({
      path: note.path,
      title: note.title.trim() || 'Untitled',
      folder: note.folder,
      updatedAt: note.updatedAt,
      pinned
    })
  }
  for (const path of pinnedPaths) {
    const note = byPath.get(path)
    if (note) push(note, true)
  }
  for (const note of live.slice().sort((a, b) => b.updatedAt - a.updatedAt)) {
    if (out.length >= max) break
    push(note, false)
  }
  return out
}

export type WidgetTaskSource = Pick<
  VaultTask,
  'id' | 'sourcePath' | 'noteTitle' | 'content' | 'due' | 'inProgress' | 'priority'
>

/**
 * Drop tasks whose note no longer exists (or sits in Trash). App-core keeps
 * `vaultTasks` fresh only while a tasks surface is on screen, so on the
 * phone a note deleted from the drawer can leave its tasks in the cache
 * until the next full scan — the widget must not show them.
 */
export function filterLiveTasks<T extends { sourcePath: string }>(
  tasks: readonly T[],
  notes: readonly Pick<NoteMeta, 'path' | 'folder'>[]
): T[] {
  const live = new Set(notes.filter((n) => n.folder !== 'trash').map((n) => n.path))
  return tasks.filter((t) => live.has(t.sourcePath))
}

/**
 * The rows for the Tasks widget from the Today bucket app-core's
 * `computeTasksRender` produces (due today, overdue, or undated — the same
 * list the Home dashboard shows), plus the counts the header needs even
 * when rows are cut off. `todayIso` is the local calendar day.
 *
 * One departure from the bucket's file order: overdue tasks lead. The
 * widget shows three to eight rows under a header that counts the overdue
 * ones, and in a vault with many undated tasks the bucket order would keep
 * every overdue row out of sight. The sort is stable, so everything else
 * keeps the app's order.
 */
export function selectWidgetTasks(
  today: readonly WidgetTaskSource[],
  overdueCount: number,
  todayIso: string,
  max = WIDGET_MAX_TASKS
): { tasks: WidgetTask[]; counts: WidgetTaskCounts } {
  const isOverdue = (t: WidgetTaskSource): boolean => typeof t.due === 'string' && t.due < todayIso
  const ordered = today.slice().sort((a, b) => Number(isOverdue(b)) - Number(isOverdue(a)))
  const tasks = ordered.slice(0, max).map(
    (t): WidgetTask => ({
      id: t.id,
      path: t.sourcePath,
      noteTitle: t.noteTitle,
      content: t.content.trim() || 'Untitled task',
      due: t.due ?? null,
      overdue: typeof t.due === 'string' && t.due < todayIso,
      inProgress: t.inProgress,
      priority: t.priority ?? null
    })
  )
  return { tasks, counts: { today: today.length, overdue: overdueCount } }
}
